src/sources/logstash.rs

```diff
@@ -0,0 +1,762 @@
+        }
+    }
+}
+
+impl From<io::Error> for DecodeError {
+    fn from(source: io::Error) -> Self {
+        DecodeError::IO { source }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum LogstashProtocolVersion {
+    V1, // 1
+    V2, // 2
+}
+
+impl From<LogstashProtocolVersion> for u8 {
+    fn from(frame_type: LogstashProtocolVersion) -> u8 {
+        use LogstashProtocolVersion::*;
+
+        match frame_type {
+            V1 => b'1',
+            V2 => b'2',
+        }
+    }
+}
+
+impl TryFrom<u8> for LogstashProtocolVersion {
+    type Error = DecodeError;
+
+    fn try_from(frame_type: u8) -> Result<LogstashProtocolVersion, DecodeError> {
+        use LogstashProtocolVersion::*;
+
+        match frame_type {
+            b'1' => Ok(V1),
+            b'2' => Ok(V2),
+            version => Err(DecodeError::UnknownProtocolVersion {
+                version: version as char,
+            }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum LogstashFrameType {
+    Ack,        // A
+    WindowSize, // W
+    Data,       // D
+    Json,       // J
+    Compressed, // C
+}
+
+impl From<LogstashFrameType> for u8 {
+    fn from(frame_type: LogstashFrameType) -> u8 {
+        use LogstashFrameType::*;
+
+        match frame_type {
+            Ack => b'A',
+            WindowSize => b'W',
+            Data => b'D',
+            Json => b'J',
+            Compressed => b'C',
+        }
+    }
+}
+
+impl TryFrom<u8> for LogstashFrameType {
+    type Error = DecodeError;
+
+    fn try_from(frame_type: u8) -> Result<LogstashFrameType, DecodeError> {
+        use LogstashFrameType::*;
+
+        match frame_type {
+            b'A' => Ok(Ack),
+            b'W' => Ok(WindowSize),
+            b'D' => Ok(Data),
+            b'J' => Ok(Json),
+            b'C' => Ok(Compressed),
+            frame_type => Err(DecodeError::UnknownFrameType {
+                frame_type: frame_type as char,
+            }),
+        }
+    }
+}
+
+/// Normalized event from logstash frame
+#[derive(Debug)]
+struct LogstashEventFrame {
+    protocol: LogstashProtocolVersion,
+    sequence_number: u32,
+    fields: BTreeMap<String, serde_json::Value>,
+}
+
+// Based on spec at: https://github.com/logstash-plugins/logstash-input-beats/blob/master/PROTOCOL.md
+// And implementation from logstash: https://github.com/logstash-plugins/logstash-input-beats/blob/27bad62a26a81fc000a9d21495b8dc7174ab63e9/src/main/java/org/logstash/beats/BeatsParser.java
+impl Decoder for LogstashDecoder {
+    type Item = LogstashEventFrame;
+    type Error = DecodeError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        // This implements a sort of simple state machine to read the frames from the wire
+        //
+        // Each matched arm with either:
+        // * Return that there is not enough data
+        // * Return an error
+        // * Read some bytes and advance the state
+        loop {
+            match self.state {
```

Comment:
Consider using `self.state = match self.state …` here, it will shorten some of the arms, assuming there aren't ownership issues. I think there would be one assign still needed as it also returns a value. You could integrate the return value in (`let (new_state, result) = match …`), but that'll probably just add complication.

Reply:
👍 this did make it a little more concise, thanks!